docs/userguide/design.md

```diff
@@ -206,35 +206,84 @@ service objects, organized according to locale. Then, if a particular locale's
+
+2. ICU Data Locating Functions, `u_setCommonData()`, `u_setDataDirectory()`, and
+`u_setAppData()`. One or more of these functions will be required when ICU is
+configured to load its data directly from files rather than taking it from the
+default data dll, and the files are not in the default location. Again, this is
+not common. See [ICU Data](icudata#icu-data-directory).
+
+3. `u_init()`. This function performs a sanity check on ICU's ability to access
+its data. It can be used to provide an early indication of configuration or
+installation problems with ICU data files. Its use is optional; it is not
+required for the correct functioning of ICU.
 
 When an application is terminating, it may optionally call the function
-`u_cleanup(void)`, which will free any heap storage that has been allocated and
-held by the ICU library. The main benefit of `u_cleanup()` occurs when using
-memory leak checking tools while debugging or testing an application. Without
-`u_cleanup()`, memory being held by the ICU library will be reported as leaks.
+`u_cleanup()`, which frees all heap storage and other system resources that
+are held internally by the ICU library. The main benefit of `u_cleanup()` occurs
+when using memory leak checking tools while testing an application. Without
+`u_cleanup()`, resources being held by ICU library will be reported as leaking.
+
+Before calling `u_cleanup()`, all ICU objects that were created by the
+application must be deleted, and all ICU services (plain C APIs) must be closed.
+
+For some platforms the configure option `--enable-auto-cleanup`, or defining
+the option `UCLN_NO_AUTO_CLEANUP` to 0, will add code which automatically cleans
+up ICU when its shared library is unloaded. See comments in `ucln_imp.h`
+
+#### C++ Static Initialization and Destruction
+
+The ICU library itself does not rely on C++ static initializers, meaning that
+applications will not encounter order-of-initialization problems from the use of
+ICU.
+
+There are, however, some significant limitations for applications that make use
+of ICU at C++ static initialization time:
 
-(For some platforms, the configure option `--enable-auto-cleanup` (or
-defining the option `UCLN_NO_AUTO_CLEANUP` to 0) will add code which
-automatically cleans up ICU when its shared library is unloaded. See comments in
-`ucln_imp.h`)
+1.  `u_setMemoryFunctions()` and the data locating functions, if needed, must
+still be called before any other use of ICU. Which includes any use during the
+construction of static objects.
+
+2.  `u_cleanup()` can only be called after all ICU objects have been destructed
+(C++) or closed (plain C APIs). Meaning that `u_cleanup()` cannot be called
+directly just before the application's `main()` function returns. It must either
+be skipped, or called from some other static destructor that runs after all use
+of ICU is done. Skipping `u_cleanup()` is not desirable; doing so will cause leak
```

Comment:
This seems confusing: Above, u_cleanup() is option; now, skipping it is not desirable.

I think what you mean is, when you use a leak checker, you do want to call u_cleanup(), you just have to find a good time & place after all of your ICU-using static objects are cleaned up.

Reply:
I agree that it's inconsistent, and am changing the earlier section to recommend the use of u_cleanup(). And revising the text here.

I'm almost to the point of saying that application code that is not compatible with leak checkers should be considered fatally broken, meaning that u_cleanup() is mandatory. And that C++ static init is so broken that we should not claim to support ICU use at static init, for lack of any reliable way to call u_cleanup(). But maybe some environments have some way to do it that I don't know about. And it will work OK for small apps, with all ICU static objects defined in a single file, which gives a well defined ordering.

I don't think it would be appropriate to write a whole treatise on C++ static init problems and solutions for this section of the user guide.